### PR TITLE
Support instance-managers for v1 and v2 volumes respectively

### PR DESCRIPTION
**Pull request by derekbit**
longhorn/longhorn#6984

### CODE REVIEW

File: controller/node_controller.go [hunk 1]
```diff
@@ -941,90 +1028,88 @@ func (nc *NodeController) syncInstanceManagers(node *longhorn.Node) error {
...
 
-	imTypes := []longhorn.InstanceManagerType{
-		longhorn.InstanceManagerTypeAllInOne,
-		longhorn.InstanceManagerTypeEngine,
+	imTypes := getInstanceManagerTypes(node)
+	backendStoreDrivers := nc.getBackendStoreDrivers(node)
+
+	// Clean up instance manager for v2 data engine if the setting is disabled
+	if _, ok := backendStoreDrivers[longhorn.InstanceManagerTypeAllInOne][longhorn.BackendStoreDriverTypeV2]; !ok {
+		if err := nc.cleanupInstanceManagerForV2DataEngine(node); err != nil {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/node_controller.go:1036]: Just instanceManager sync operations together. Any potential issue?

File: controller/node_controller.go [hunk 2]
```diff
@@ -933,6 +933,93 @@ func (nc *NodeController) syncNodeStatus(pod *corev1.Pod, node *longhorn.Node) e
...
+	}
+
+	if len(node.Spec.Disks) != 0 {
+		imTypes = append(imTypes, longhorn.InstanceManagerTypeReplica)
+	}
+
+	return imTypes
+}
+
+func (nc *NodeController) cleanUpAllReplicaManagers(node *longhorn.Node) error {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/node_controller.go]: Updated. Thank you.

File: upgrade/v15xto160/upgrade.go [hunk 3]
```diff
@@ -70,6 +74,34 @@ func upgradeVolumes(namespace string, lhClient *lhclientset.Clientset, resourceM
...
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn instanceManagers during the instanceManager upgrade")
+	}
+
+	for _, im := range imMap {
+		labels := im.Labels
+		if labels != nil {
+			labels[types.GetLonghornLabelKey(types.LonghornLabelBackendStoreDriver)] = string(longhorn.BackendStoreDriverTypeV1)
+		}
+		im.Labels = labels
```
[the remark this answers is not in the document]
Comment (derekbit) [reply upgrade/v15xto160/upgrade.go]: Updated. Thank you.

File: webhook/resources/instancemanager/mutator.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/pkg/errors"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
```
[the remark this answers is not in the document]
Comment (derekbit) [reply webhook/resources/instancemanager/mutator.go]: Updated. Thank you.

File: controller/node_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (c3y1huang) [reply controller/node_controller.go:1036]: In my opinion, it would make more sense for the setting controller or the instance-manager controller to handle the cleanup when the Setting CR is updated. The node controller doesn't appear to have a direct link to the instance-manager pods or the v2 engine setting, which makes the implementation less straightforward.

[the remark this answers is not in the document]
Comment (derekbit) [reply controller/node_controller.go]: Understand. Addressed it. Thank you.

File: controller/monitor/disk_monitor.go [hunk 5]
```diff
@@ -278,14 +280,14 @@ func (m *NodeMonitor) collectDiskData(node *longhorn.Node) map[string]*Collected
 			continue
 		}
 
-		orphanedReplicaDirectoryNames, err = m.getOrphanedReplicaInstanceNames(disk.Type, node, diskName, diskConfig.DiskUUID, disk.Path, replicaInstanceNames)
+		orphanedReplicaInstanceNames, err = m.getOrphanedReplicaInstanceNames(disk.Type, node, diskName, diskConfig.DiskUUID, disk.Path, replicaInstanceNames)
```
Comment (c3y1huang) [inline controller/monitor/disk_monitor.go:283]: This looks like it still returns orphaned replica directory names. Why the change in the variable name?

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: controller/monitor/disk_monitor.go (hunk 5 above)
Comment (derekbit) [reply controller/monitor/disk_monitor.go:283]: It returnn replica lvol bdev for v2 volunes, so using orphanedReplicaInstanceNames is better.

File: controller/backing_image_controller.go [hunk 6]
```diff
@@ -108,8 +108,8 @@ func (bic *BackingImageController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	defer bic.queue.ShutDown()
 
-	logrus.Info("Starting Longhorn Backing Image controller")
-	defer logrus.Info("Shut down Longhorn Backing Image controller")
+	bic.logger.Info("Starting Longhorn Backing Image controller")
+	defer bic.logger.Info("Shut down Longhorn Backing Image controller")
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/backing_image_controller.go:112]: Seems correct for the past tense.

File: controller/controller_test.go [hunk 7]
```diff
@@ -610,12 +610,12 @@ func (s *TestSuite) TestIsSameGuaranteedCPURequirement(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(IsSameGuaranteedCPURequirement(a, b), Equals, true)
 
-	b.Requests[corev1.ResourceCPU], err = resource.ParseQuantity("250m")
+	b.Requests[corev1.ResourceCPU], _ = resource.ParseQuantity("250m")
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/controller_test.go]: Added mssing `c.Assert(err, IsNil)`, thank you.

File: controller/monitor/disk_monitor.go [hunk 8]
```diff
@@ -138,40 +138,68 @@ func (m *NodeMonitor) run(value interface{}) error {
...
+	}
+	if v2DataEngineEnabled {
+		backendStoreDrivers = append(backendStoreDrivers, longhorn.BackendStoreDriverTypeV2)
 	}
 
-	return engineapi.NewDiskServiceClient(im, m.logger)
+	return backendStoreDrivers
+}
+
+func (m *NodeMonitor) newDiskServiceClient(node *longhorn.Node) (map[longhorn.BackendStoreDriverType]*engineapi.DiskService, error) {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/monitor/disk_monitor.go]: Updated, thank you.

File: controller/snapshot_controller.go [hunk 9]
```diff
@@ -304,7 +304,6 @@ func (sc *SnapshotController) handleErr(err error, key interface{}) {
 	log := sc.logger.WithField("Snapshot", key)
 	handleReconcileErrorLogging(log, err, "Failed to sync Longhorn snapshot")
 	sc.queue.AddRateLimited(key)
-	return
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/snapshot_controller.go:307]: Addressed. Thank you.

**Comment by derekbit** [comment]
@c3y1huang Would you be able to review the PR? Thank you.

**Comment by derekbit** [comment]
Closed and reviewed in https://github.com/longhorn/longhorn-manager/pull/2371/commits
